jumanji/environments/routing/rware/env.py

```diff
@@ -0,0 +1,527 @@
+        for every successful delivery of a requested shelf to the goal position.
+
+    - episode termination:
+        - The number of steps is greater than the limit.
+        - Any agent selects an action which causes two agents to collide.
+
+    - state: State
+        - grid: an array representing the warehouse floor as a 2D grid with two separate channels
+            one for the agents, and one for the shelves
+        - agents: a pytree of Agent type with per agent leaves: [position, direction, is_carrying]
+        - shelves: a pytree of Shelf type with per shelf leaves: [position, is_requested]
+        - request_queue: the queue of requested shelves (by ID).
+        - step_count: an integer representing the current step of the episode.
+        - key: a pseudorandom number generator key.
+
+    [1] Papoudakis et al., Benchmarking Multi-Agent Deep Reinforcement Learning Algorithms
+        in Cooperative Tasks (2021)
+
+    ```python
+    from jumanji.environments import Rware
+    env = Rware()
+    key = jax.random.PRNGKey(0)
+    state, timestep = jax.jit(env.reset)(key)
+    env.render(state)
+    action = env.action_spec().generate_value()
+    state, timestep = jax.jit(env.step)(state, action)
+    env.render(state)
+    ```
+    """
+
+    def __init__(
+        self,
+        generator: Optional[Generator] = None,
+        time_limit: int = 500,
+        viewer: Optional[Viewer[State]] = None,
+    ):
+        """Instantiates an `Rware` environment.
+
+        Args:
+            generator: callable to instantiate environment instances.
+                Defaults to `RandomGenerator` with parameters:
+                `shelf_rows = 1`,
+                `shelf_columns = 3`,
+                `column_height = 8`,
+                `num_agents = 2`,
+                `sensor_range = 1`,
+                `request_queue_size = 4`.
+            time_limit: the maximum step limit allowed within the environment.
+                Defaults to 500.
+        """
+
+        self._generator = generator or RandomGenerator(
+            shelf_rows=1,
+            shelf_columns=3,
+            column_height=8,
+            num_agents=2,
+            sensor_range=1,
+            request_queue_size=4,
+        )
+        self.goals: List[Tuple[int, int]] = []
+        self.grid_size = self._generator.grid_size
+        self.request_queue_size = self._generator.request_queue_size
+
+        self.num_agents = self._generator.num_agents
+        self.sensor_range = self._generator.sensor_range
+        self.highways = self._generator.highways
+        self.shelf_ids = self._generator.shelf_ids
+        self.not_in_queue_size = self._generator.not_in_queue_size
+
+        self.agent_ids = jnp.arange(self.num_agents)
+        self.directions = jnp.array([d.value for d in Direction])
+        self.num_obs_features = utils.calculate_num_observation_features(
+            self.sensor_range
+        )
+        self.goals = self._generator.goals
```

Comment:
Think it would be cleaner if we remove line 154 and then moved this up to line 163, just below where we're caching everything else from the generator